drivers/led/led_gpio.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <drivers/gpio.h>
+#include <drivers/led.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(led_gpio, CONFIG_LED_GPIO_LOG_LEVEL);
+
+#define DT_DRV_COMPAT gpio_leds
+#define LED_DATA(node_id)                      \
+{                                              \
+	{                                      \
+		DT_GPIO_LABEL(node_id, gpios), \
+		DT_GPIO_PIN(node_id, gpios),   \
+		DT_GPIO_FLAGS(node_id, gpios)  \
+	},                                     \
+	DT_LABEL(node_id),                     \
+	DT_PROP(node_id, id)                   \
+},
+#define LED_SW_BLINK BIT(0)
+#define LED_BLINK_DISABLE BIT(1)
+
+struct gpio_cfg {
+	const char *port;
+	uint8_t pin;
+	uint8_t polarity;
+};
+
+struct led_dts_cfg {
+	struct gpio_cfg gpio;
+	const char *label;
+	uint32_t id;
+};
+
+struct led_gpio_data {
+	struct device *gpio_dev;
+	const struct led_dts_cfg *cfg;
+#if defined(CONFIG_LED_SOFTWARE_BLINK)
+	struct k_timer timer;
+	atomic_t flags;
+	uint32_t blink_delay_on;
+	uint32_t blink_delay_off;
+	bool next_state;
+#endif
+	bool state;
+};
+
+struct led_gpio_cfg {
+	const struct led_dts_cfg *led;
+	uint32_t led_cnt;
+};
+
+static inline struct led_gpio_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct led_gpio_cfg *get_dev_config(struct device *dev)
+{
+	return dev->config_info;
+}
+
+static struct led_gpio_data *led_parameters_get(struct device *dev, uint32_t led)
+{
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		if (data[i].cfg->id == led) {
+			return &data[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int gpio_led_set_no_delay(struct led_gpio_data *data, bool state)
+{
+	int err;
+	uint32_t pin_value;
+
+	data->state = state;
+
+	if (data->cfg->gpio.polarity & GPIO_OUTPUT_ACTIVE) {
```

Comment:
Remove all this.  It should be handled by `gpio_pin_configure()`.

Reply:
Removed